File: py/sort/array/BucketSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class BucketSort:
    def sort(self, nums: List[int]) -> None:
        maxValue, minValue = max(nums), min(nums)

        size = (maxValue - minValue) // 10 + 1
        buckets = [[] for _ in range(size)]
        for value in nums:
            buckets[(value - minValue) // 10].append(value)

        index = 0
        for bucket in buckets:
            bucket.sort()

            for value in bucket:
                nums[index] = value
                index += 1

File: cpp/sort/array/BucketSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class BucketSort {
public:
    void sort(vector<int> &nums) {
        int maxValue = *max_element(begin(nums), end(nums));
        int minValue = *min_element(begin(nums), end(nums));

        int size = (maxValue - minValue) / 10 + 1;
        vector<vector<int>> buckets(size, vector<int>(0));
        for (int value : nums) {
            buckets[(value - minValue) / 10].emplace_back(value);
        }

        int index = 0;
        for (vector<int> &bucket : buckets) {
            std::sort(bucket.begin(), bucket.end());

            for (int value : bucket) {
                nums[index] = value;
                ++index;
            }
        }
    }
};
